src/libopensc/card-idprime.c

```diff
@@ -405,11 +707,26 @@ static int idprime_fill_prkey_info(list_t *list, idprime_object_t **entry, sc_pk
 	prkey_info->id.value[0] = ((*entry)->fd >> 8) & 0xff;
 	prkey_info->id.value[1] = (*entry)->fd & 0xff;
 	prkey_info->id.len = 2;
-	prkey_info->key_reference = (*entry)->key_reference;
+	if ((*entry)->valid_key_ref)
+		prkey_info->key_reference = (*entry)->key_reference;
+	else
+		prkey_info->key_reference = -1;
 	*entry = list_iterator_next(list);
 	return SC_SUCCESS;
 }
 
+/* get PIN id of the current object on the list */
+static int idprime_get_pin_id(list_t *list, idprime_object_t **entry, char **pin_id)
+{
+	if (pin_id == NULL || entry == NULL) {
+		return SC_ERROR_INVALID_ARGUMENTS;
+	}
+	*pin_id = "11"; // normal PIN id
```

Comment:
maybe you want to make `pin_id` a `const char **`  to indicate that the returned memory does not need to be `free`d.